### DISCUSSION TITLE
ringbuf: Use SetDeadline to consume data ?

### DISCUSSION THREAD

**Discussion by RonFed** [cilium/ebpf#1108] [Q&A] [answered]
Hi,
I'm interested in setting a timeout for Read in ringbuf which will consume records if possible.
Currently it seems like the SetDeadline returns an error upon timeout and not trying to read from the buffer.

The use case for this is as follow:
From the eBPF program I want to wake the user code once every X submissions to the ringbuf (https://patchwork.ozlabs.org/project/netdev/patch/20200529075424.3139988-5-andriin@fb.com/ explains the benefits of this ''sampled' approach). However, I want to be able to consume the data from the user side in case it didn't get a signal after T seconds - and consume all the available data.

Is this possible using the current ringbuf API ?

**Comment by mythi** [comment]
It's been a while since I looked into this but IIRC that use case was [discussed in the original ringbuf review](https://github.com/cilium/ebpf/pull/318#discussion_r694746718) but it was left out from the initial version.

Perhaps SetDeadline err could indeed check if data is available and set `r.haveData`.

**Reply by RonFed** [reply]
Another approach might be to add an option for a non-blocking Read using the following code from ringbuf/reader.go
```
                 for {
			err := readRecord(r.ring, rec, r.header)
			if err == errBusy || err == errDiscard {
				continue
			}
			if err == errEOR {
				r.haveData = false
				break
			}

			return err
		}
```

The user can use this non-blocking read after a timeout to consume the data if present

**Comment by lmb** [comment]
Here is what I think you want to do: set BPF_RB_NO_WAKEUP on all but some events from the BPF side. You're turning of epoll notifications but still want to be able to flush the buffer periodically, since you're not guaranteed a next event. Next time please describe your goals directly in addition to linking to the patchset, it would've saved me some time reading.

I think that Mykko is right: we could set `haveData = true` if we encounter `ErrDeadlineExceeded` to force a single load from the ringbuffer. That allows supporting the behaviour without additional API

**Reply by RonFed** [reply]
Right, you described my goal accurately.

**Reply by lmb** [reply]
Great! Are you up for implementing it? The change itself is probably trivial. For testing you'd have to add a new program which does a RB_NO_WAKEUP write.

**Reply by RonFed** [reply]
Sure, I'll start working on it

**Comment by RonFed** [comment, marked as answer]
PR is at https://github.com/cilium/ebpf/pull/1111 @lmb @mythi
